test/metabase/models/params/custom_values_test.clj

```diff
@@ -0,0 +1,104 @@
+(ns metabase.models.params.custom-values-test
+  (:require
+    [clojure.test :refer :all]
+    [metabase.models :refer [Card]]
+    [metabase.models.params.custom-values :as custom-values]
+    [metabase.test :as mt]
+    [toucan.db :as db]))
+
+(defn- find-field-ref
+  [id-or-name result-metadata]
+  (let [selector (if (string? id-or-name)
+                   :name
+                   :id)]
+    (-> (filter #(= id-or-name (selector %)) result-metadata)
+        first
+        :field_ref)))
+
+(deftest with-mbql-card-test
```

Comment:
Can we test this works with a field that has a `:binning` strategy, and also a field with `:temporal-unit` bucketing?

Reply:
added tests for temporal-unit bucketing.

With binning, things get a bit weird because I'm not sure if field-ref from FE will contain the binning info or not (the full `[:field 42 {:binning {:strategy ... }}]`

I'll just throw an error for now: https://metaboat.slack.com/archives/C04DAKKDMFT/p1671632554048439?thread_ts=1671611125.666239&cid=C04DAKKDMFT.

Will revisit this later.

Reply:
Added an item to the nice to have in this list: https://github.com/metabase/metabase/issues/26873

Reply:
added tests for temporal bucketing.

For binning it's not clear to me whether the binning info will be included in the field_ref.

if it's already included then it'll work with no problems.
I added a todo to our milestone to review this later: https://github.com/metabase/metabase/issues/26873

Currently I'll just throw an error if users select binning column.

Reply:
When you do revisit it, I still think temporal-unit should be explicitly absent from the MBQL query too.

The test passes because applying temporal-unit on a field in the MBQL query returns the same results as querying the field with no temporal-unit. But it's only because the field in the source card's query has had temporal bucketing applied previously, and temporal bucketing is idempotent. It smells a bit magic. Ideally be doing is coming up with a consistent way to query fields / aggregations without applying temporal bucketing or binning or _anything_ like that.

Reply:
Isn't the temporal-unit already excluded?
Currently only `:source-field :join-alias :base-type` is selected. so the temporal-unit returned by the card will be what it set in the original card.

Reply:
Oh, I didn't realise you were only selecting those. I thought you were throwing an error on binning because you were defferring how to deal with those extra field options. If you're already selecting only :source-field, :join-alias, and :base-type, then isn't binning already dealt with and you don't need to throw an exception? It seems like you're already handling binning fine.

Reply:
When I try from the UI, it includes a `binning` map.

Even though I tried to call the same query but remove the `binning` map and it returns the sam result.

But I'm not sure how important the binning map is. So let just ignore it for now and review this later.
<img width="2126" alt="Screenshot 2022-12-22 at 16 17 37" src="https://user-images.githubusercontent.com/25661381/209100794-3040172f-8f6a-4d85-b6f2-93d25192eb4e.png">

Reply:
Oh, of course. It makes sense to me now. Binning is idempotent too, like temporal unit. If you bin a column into 10 bins, and bin that again, and again etc it will return the same result as just binning once. So binning and temporal-unit don't need to be stripped out to produce the correct results. But it smells. The MBQL query shouldn't have any options that apply additional operations other than just selecting the field.

Reply:
Sure, that wouldn't be idempotent. What I meant to say is not "binning is idempotent" in general, but "binning is idempotent when the binning parameters are the same", i.e.
<img width="81" alt="image" src="https://user-images.githubusercontent.com/39073188/209124373-b7822750-dac4-4590-b2c6-30d4888b5e5d.png">

Same goes for temporal bucketing.